### PR TITLE
Add ability to select system as a reference

### PR DESCRIPTION
**Pull request by johnsonm325**

### CODE REVIEW

**Comment by johnsonm325** [comment]
Just remembered we'll have to wait for https://github.com/RedHatInsights/drift-backend/pull/152

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 1]
```diff
@@ -91,7 +110,11 @@ export class DriftTable extends Component {
         }
         /*eslint-enable camelcase*/
 
-        return baselines.concat(fullHistoricalSystemList);
+        masterList =
+            this.referenceId
+                ? this.moveSystem(baselines.concat(fullHistoricalSystemList))
+                : baselines.concat(fullHistoricalSystemList);
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js]: This performs a concat in both cases so I would opt to simplify this and move that up a line something like:

```
masterList = baselines.concat(fullHistoricalSystemLIst);

if (this.referenceId) {
  masterList = this.moveSystem(masterList);
}
```

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 2]
```diff
@@ -38,13 +40,31 @@ export class DriftTable extends Component {
         await window.insights.chrome.auth.getUser();
         const { fetchCompare } = this.props;
 
-        fetchCompare(this.systemIds, this.baselineIds, this.HSPIds);
+        fetchCompare(this.systemIds, this.baselineIds, this.HSPIds, this.referenceId);
+    }
+
+    moveSystem = (systems) => {
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js]: We are generally running into this problem, where we are dealing with baselines, systems and historical profiles. So the use of system nomenclature can get confusing. Naming ideas then `shiftReferenceToFront` and pass in `items` or `masterList`

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 3]
```diff
@@ -38,13 +40,31 @@ export class DriftTable extends Component {
...
+
+        systems.forEach(function(system) {
+            if (referenceId === system.id) {
+                index = systems.indexOf(system);
+                systemToMove = system;
+            }
+        });
+
+        systems.splice(index, 1);
+        systems.unshift(systemToMove);
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js]: I think you can simplify all of this logic a bit using some other builtins:

```
index = systems.findIndex(function (item) {
    return referenceId === system.id;
});

systemToMove = systems.splice(index, 1);
systems.unshift(systemToMove);
```

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 4]
```diff
@@ -183,11 +216,23 @@ export class DriftTable extends Component {
...
+                        system.meets_reference_expectations || system.meets_reference_expectations === undefined
+                            ? 'comparison-cell'
+                            : 'highlight comparison-cell different-fact-cell' }
+                    >
+                        { system.value === null ? 'No Data' : system.value }
+                    </td>
+                );
+            } else {
+                row.push(
+                    <td className={ fact.state === 'DIFFERENT' ? 'highlight comparison-cell' : 'comparison-cell' }>
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js]: Looks like the only difference between the two `row.push` statements is how this className is set. I think then you can extract that to logic thats wrapped in an if-else and call row.push only once.

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 5]
```diff
@@ -256,6 +301,8 @@ export class DriftTable extends Component {
     renderSystemHeaders() {
         let row = [];
         let typeIcon = '';
+        let referenceId = this.referenceId;
+        let updateReferenceId = this.updateReferenceId;
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js]: I sometimes find casting instance variables to local variables unnecessary. The additional use of `this.` in this function to reference those instance variables is easier to parse when reading the code.

File: src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    render() {
+        const { updateReferenceId, id } = this.props;
+        const { isReference } = this.state;
+
+        return (
+            isReference
+                ? <StarIcon
+                    className='reference-selector pointer'
+                    onClick={ () => updateReferenceId(undefined) }
+                />
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js]: Should this be indented one more step to align with the `<StarIcon` ?

File: src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                : <OutlinedStarIcon
+                    className='reference-selector pointer'
+                    onClick={ () => updateReferenceId(id) }
+                />
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js]: Should this be indented one more step to align with the `<StarIcon` ?

File: src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    render() {
+        const { updateReferenceId, id } = this.props;
+        const { isReference } = this.state;
+
+        return (
+            isReference
+                ? <StarIcon
+                    className='reference-selector pointer'
+                    onClick={ () => updateReferenceId(undefined) }
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js]: You shouldn't need to call undefined explicitly here.

File: src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js (hunk 6 above)
Comment (johnsonm325) [reply src/SmartComponents/DriftPage/DriftTable/ReferenceSelector/ReferenceSelector.js:23]: No, the `<StarIcon` isn't indented, it's just a space between the `?` and `<StarIcon`

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 9]
```diff
@@ -177,17 +210,28 @@ export class DriftTable extends Component {
...
+                } else {
+                    className = 'comparison-cell';
+                }
+            } else {
+                if (fact.state === 'DIFFERENT') {
+                    className = 'highlight comparison-cell';
+                } else {
+                    className = 'comparison-cell';
+                }
+            }
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js:232]: Note that the code adds `comparison-cell` in all cases. The `else` clauses could be dropped by starting with that:

```
let className = ['comparison-cell']

if (fact.state === 'DIFFERENT') {
     className.push('highlight');

     if (this.referenceId) {
         className.push('different-fact-cell');
}

row.push(<td className={ className,join() }>{ system.value === null ? 'No Data' : system.value }</td>);
```

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js [hunk 10]
```diff
@@ -38,13 +40,29 @@ export class DriftTable extends Component {
         await window.insights.chrome.auth.getUser();
         const { fetchCompare } = this.props;
 
-        fetchCompare(this.systemIds, this.baselineIds, this.HSPIds);
+        fetchCompare(this.systemIds, this.baselineIds, this.HSPIds, this.referenceId);
+    }
+
+    shiftReferenceToFront = (masterList) => {
+        let referenceId = this.referenceId;
```
Comment (ehelms) [inline src/SmartComponents/DriftPage/DriftTable/DriftTable.js:47]: For me, this cast is not necessary and `this.referenceId` can be used in the body of the function.

File: src/SmartComponents/DriftPage/DriftTable/DriftTable.js (hunk 9 above)
Comment (johnsonm325) [reply src/SmartComponents/DriftPage/DriftTable/DriftTable.js:232]: Great change!

**Comment by johnsonm325** [comment]
Your logic change for `className` above didn't work like I thought it would, so I had to revert to something similar to what I already had.

**Pull request merged**